openmetadata-service/src/main/java/org/openmetadata/service/apps/bundles/changeEvent/AbstractEventConsumer.java

```diff
@@ -0,0 +1,316 @@
+
+  }
+
+  @Override
+  public void handleFailedEvent(EventPublisherException ex) {
+    LOG.debug(
+        "Failed in for Publisher : {} , Change Event : {} ",
+        eventSubscription.getName(),
+        ex.getChangeEvent());
+
+    ChangeEvent event = ex.getChangeEvent();
+
+    // Update Failed Event with details
+    FailedEvent failedEvent =
+        new FailedEvent()
+            .withChangeEvent(ex.getChangeEvent())
+            .withRetriesLeft(eventSubscription.getRetries())
+            .withTimestamp(System.currentTimeMillis());
+
+    if (eventSubscription.getRetries() == 0) {
+      Entity.getCollectionDAO()
+          .eventSubscriptionDAO()
+          .upsertFailedEvent(
+              eventSubscription.getId().toString(),
+              String.format("%s-%s", FAILED_EVENT_EXTENSION, event.getId()),
+              JsonUtils.pojoToJson(failedEvent));
+    } else {
+      // Check in Qtz Map
+      Set<FailedEvent> failedEventsList =
+          JsonUtils.convertValue(
+              jobDetail.getJobDataMap().get(FAILED_EVENT_EXTENSION), new TypeReference<>() {});
+      if (failedEventsList == null) {
+        failedEventsList = new HashSet<>();
+      } else {
+        // Remove exising change event
+        boolean removeChangeEvent =
+            failedEventsList.removeIf(
+                failedEvent1 -> {
+                  if (failedEvent1
+                      .getChangeEvent()
+                      .getId()
+                      .equals(failedEvent.getChangeEvent().getId())) {
+                    failedEvent.withRetriesLeft(failedEvent1.getRetriesLeft());
+                    return true;
+                  }
+                  return false;
+                });
+
+        if (removeChangeEvent) {
+          if (failedEvent.getRetriesLeft() == 0) {
+            // If the Retries are exhausted, then remove the Event from the List to DLQ
+            failedEvent.withRetriesLeft(0);
+          } else {
+            failedEvent.withRetriesLeft(failedEvent.getRetriesLeft() - 1);
+          }
+        }
+      }
+      failedEventsList.add(failedEvent);
+      jobDetail.getJobDataMap().put(FAILED_EVENT_EXTENSION, failedEventsList);
+      Entity.getCollectionDAO()
+          .eventSubscriptionDAO()
+          .upsertFailedEvent(
+              eventSubscription.getId().toString(),
+              String.format("%s-%s", FAILED_EVENT_EXTENSION, event.getId()),
+              JsonUtils.pojoToJson(failedEvent));
+    }
+  }
+
+  private int loadInitialOffset(JobExecutionContext context) {
+    EventSubscriptionOffset jobStoredOffset =
+        (EventSubscriptionOffset) jobDetail.getJobDataMap().get(ALERT_OFFSET_KEY);
+    // If the Job Data Map has the latest offset, use it
+    if (jobStoredOffset != null) {
+      return jobStoredOffset.getOffset();
+    } else {
+      EventSubscriptionOffset eventSubscriptionOffset =
+          getInitialAlertOffsetFromDb(eventSubscription.getId());
```

Comment:
we shouldn't be defaulting to initial offset.
Lets say I am creating new alert, which means there is no offset stored for this alert but we have 1000s of rows in change_event table. if we go to initial offset this will pull all of those and publish it to slack or email. 
This becomes a spam event

Reply:
@harshach this is not the initial offset , but rather the latest count of rows in changeEvent!